### PR TITLE
feat(indexer): Implement cursor based pagination

### PR DESCRIPTION
**Pull request by roninjin10**
- Implement pagination in API

### CODE REVIEW

**Comment by roninjin10** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #6768** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6768" target="_blank">Graphite</a>  👈
    * **PR #6726** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6726" target="_blank">Graphite</a> 
      * **PR #6799** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6799" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6768?utm_source=stack-comment).

File: indexer/database/bridge_transfers.go [hunk 1]
```diff
@@ -183,7 +189,13 @@ func (db *bridgeTransfersDB) L2BridgeWithdrawalByCrossDomainMessengerNonce(nonce
...
 // L2BridgeDepositsByAddress retrieves a list of deposits intiated by the specified address, coupled with the L1/L2 transaction hashes
 // that complete the bridge transaction. The hashes that correspond to with the Bedrock multistep withdrawal process are also surfaced
-func (db *bridgeTransfersDB) L2BridgeWithdrawalsByAddress(address common.Address) ([]*L2BridgeWithdrawalWithTransactionHashes, error) {
+func (db *bridgeTransfersDB) L2BridgeWithdrawalsByAddress(address common.Address, cursor *string, limit *int) (*L2BridgeWithdrawalsResponse, error) {
+	// TODO add desc asc sortDirection
+	defaultLimit := 100
+	if limit == nil || *limit <= 0 {
+		limit = &defaultLimit
+	}
+
```
Comment (roninjin10) [inline indexer/database/bridge_transfers.go:234]: meh lost commits somehow

File: indexer/database/bridge_transfers.go [hunk 2]
```diff
@@ -54,11 +55,11 @@ type L2BridgeWithdrawalWithTransactionHashes struct {
 type BridgeTransfersView interface {
 	L1BridgeDeposit(common.Hash) (*L1BridgeDeposit, error)
 	L1BridgeDepositByCrossDomainMessengerNonce(*big.Int) (*L1BridgeDeposit, error)
-	L1BridgeDepositsByAddress(common.Address) ([]*L1BridgeDepositWithTransactionHashes, error)
+	L1BridgeDepositsByAddress(common.Address, *string, *int) (*L1BridgeDepositsResponse, error)
```
Comment (roninjin10) [inline indexer/database/bridge_transfers.go]: Is this the best way to do optional values?

File: indexer/database/bridge_transfers.go [hunk 3]
```diff
@@ -141,7 +156,24 @@ l1_transaction_deposits.l2_transaction_hash`)
...
 	}
 
-	return deposits, nil
+	hasNextPage := false
+	if len(deposits) > *limit {
+		hasNextPage = true
+		deposits = deposits[:*limit]
+	}
+
+	nextCursor := ""
```
Comment (roninjin10) [inline indexer/database/bridge_transfers.go:170]: should this be a pointer to represent optional or does representing it with "" make sense?

File: indexer/database/bridge_transfers.go (hunk 1 above)
Comment (roninjin10) [reply indexer/database/bridge_transfers.go:234]: Found them

File: indexer/migrations/20230523_create_schema.sql [hunk 4]
```diff
@@ -199,6 +199,7 @@ CREATE TABLE IF NOT EXISTS l2_bridge_messages(
 
 -- StandardBridge
 CREATE TABLE IF NOT EXISTS l1_bridge_deposits (
+    -- TODO rename me transaction_hash
```
Comment (roninjin10) [inline indexer/migrations/20230523_create_schema.sql]:
```suggestion
```
Not bothering for now maybe in cleanup phase

File: indexer/migrations/20230523_create_schema.sql [hunk 5]
```diff
@@ -215,6 +216,7 @@ CREATE TABLE IF NOT EXISTS l1_bridge_deposits (
     timestamp        INTEGER NOT NULL CHECK (timestamp > 0)
 );
 CREATE TABLE IF NOT EXISTS l2_bridge_withdrawals (
+    -- TODO rename me transaction_hash
```
Comment (roninjin10) [inline indexer/migrations/20230523_create_schema.sql]:
```suggestion
```

File: indexer/database/bridge_transfers.go [hunk 6]
```diff
@@ -129,8 +141,11 @@ l1_transaction_deposits.l2_transaction_hash`)
 	depositsQuery = depositsQuery.Joins("INNER JOIN l1_transaction_deposits ON l1_bridge_deposits.transaction_source_hash = l1_transaction_deposits.source_hash")
 	depositsQuery = depositsQuery.Joins("INNER JOIN l1_contract_events ON l1_transaction_deposits.initiated_l1_event_guid = l1_contract_events.guid")
 
-	// add in cursoring options
-	filteredQuery := depositsQuery.Where(&Transaction{FromAddress: address}).Order("l1_bridge_deposits.timestamp DESC").Limit(100)
+	if cursor != "" {
```
Comment (hamdiallam) [inline indexer/database/bridge_transfers.go:149]: I think we need to make sure this deposit is ordered by timestamp to make sure this works right. Realized i forgot to add the ordering clause to these.

That way the cursor also doesn't have to be returned in the response. The caller simply knows that the last entry is the cursor to pass into the next page

**Review by hamdiallam — APPROVED** [review]

File: indexer/database/bridge_transfers.go (hunk 6 above)
Comment (roninjin10) [reply indexer/database/bridge_transfers.go:149]: Will change it to timestamp in it's own pr but is block_number better?

File: indexer/database/bridge_transfers.go [hunk 7]
```diff
@@ -117,19 +118,34 @@ func (db *bridgeTransfersDB) L1BridgeDepositByCrossDomainMessengerNonce(nonce *b
...
+type L1BridgeDepositsResponse struct {
+	Deposits    []*L1BridgeDepositWithTransactionHashes
+	Cursor      string
+	HasNextPage bool
+}
+
 // L1BridgeDepositsByAddress retrieves a list of deposits intiated by the specified address, coupled with the L1/L2 transaction
 // hashes that complete the bridge transaction.
-func (db *bridgeTransfersDB) L1BridgeDepositsByAddress(address common.Address) ([]*L1BridgeDepositWithTransactionHashes, error) {
+func (db *bridgeTransfersDB) L1BridgeDepositsByAddress(address common.Address, cursor string, limit int) (*L1BridgeDepositsResponse, error) {
```
Comment (roninjin10) [inline indexer/database/bridge_transfers.go:134]:
```suggestion
// FIXME I'm not ordered by block_number in any way atm
func (db *bridgeTransfersDB) L1BridgeDepositsByAddress(address common.Address, cursor string, limit int) (*L1BridgeDepositsResponse, error) {
```

**Pull request merged**